Port the cryptocell 310 cmac driver

RonEld opened the pull request:
### Description

<!--
    Required
    Add here detailed changes summary, testing results, dependencies
    Good example: https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html (Pull request template)
-->
Add support for CC310 CMAC driver returning
`MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED` for key size other than 128 bits,
and for crypto algorithms other than AES( e.g. DES).
According to the generated map file, using GCC_ARM:
ROM: 640 bytes smaller
RAM: No Change
Benchmark information, built with GCC_ARM:
~~~
  AES-CMAC-128             :       7255 KB/s                                                                                               
  AES-CMAC-192             :  Feature unsupported                                                                                          
  AES-CMAC-256             :  Feature unsupported                                                                                          
  AES-CMAC-PRF-128         :  Feature unsupported 
~~~
### Pull request type

<!--
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [ ] Fix
    [ ] Refactor
    [ x] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

### Reviewers

<!--
    Optional
    Request additional reviewers with @username
-->
@ARMmbed/mbed-os-crypto 

### Release Notes

<!--
    Optional
    In case of breaking changes, functionality changes or refactors, please add release notes here. 
    For more information, please see [the contributing guidelines](https://os.mbed.com/docs/mbed-os/latest/contributing/workflow.html#pull-request-types).
-->
This is a target update, adding hw accelerated CMAC, however, it can also be considered as a breaking change, as it removes support for key sizes other than 128 bit keys, and removes support for DES CMAC.

Commit:
Port the cryptocell 310 cmac driver

Add support for CC310 CMAC driver returning
`MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED` for key size other than 128 bits,
and for crypto algorithms other than AES( e.g. DES).

Patater commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+            cmac_ctx->CC_keySizeInBytes = ( keybits / 8 );
+            memcpy( cmac_ctx->CC_Key, key, cmac_ctx->CC_keySizeInBytes );
+        }
+        break;
```
suggest removing the {} or putting break inside the {}

andresag01 replied:
I would advocate for removing all the brackets as the cases in this switch statement are very simple.

Patater commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.h:
```diff
@@ -0,0 +1,50 @@
+/*
+ *  aes_alt.h
```
This should say cmac_alt.h here

RonEld replied:
yes, a copy\paste error

Patater commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+        goto exit;
+    }
+
+    if( SaSi_AesFinish( &ctx.cmac_ctx->CC_Context, ilen, ( uint8_t * ) input,
```
Why not call `mbedtls_cipher_cmac_update()` and `mbedtls_cipher_cmac_finish()?

If AES_ALT is set, won't the software CMAC implementation use it?

RonEld replied:
It can be done, but it adds additional redundant code. In this API, it's a one call CMAC functionality, without storing `unprocessed_block`, and call directly to the CC AES driver to calculate the CMAC.
The alternative implementation of `mbedtls_cipher_cmac_update()` and `mbedtls_cipher_cmac_finish()` has support for the unprocessed data, to support sequential calls with sizes that are not multiple of AES  block size

Patater replied:
OK

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+     */
+    cmac_ctx = mbedtls_calloc( 1, sizeof( mbedtls_cmac_context_t ) );
+    if( cmac_ctx == NULL )
+        return( MBEDTLS_ERR_CIPHER_ALLOC_FAILED );
```
Is it necessary to allocate this before making the remaining `keybits` check? Perhaps it would be more optimal to move the allocation into the `case 128:` or move the switch check before this allocation.

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+    }
+
+    ctx->cmac_ctx = cmac_ctx;
+    return 0;
```
Please add brackets around the argument to return

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+    int ret = 0;
+    SaSiAesUserKeyData_t CC_KeyData;
+    if( SaSi_AesInit( &cmac_ctx->CC_Context, SASI_AES_ENCRYPT,
+                      SASI_AES_MODE_CMAC, SASI_AES_PADDING_NONE) != 0 )
```
Style: PLease add space before `)`

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+    }
+
+    CC_KeyData.pKey = cmac_ctx->CC_Key;
+    CC_KeyData.keySize = cmac_ctx->CC_keySizeInBytes;
```
If I understand correctly, this is just a temporary struct with a little bit of metadata a pointer to the key. Why not just make that struct part of the mbedtls_cmac_context_t? Does SaSiAesUserKeyData_t contain a lot more stuff that we do not want?

RonEld replied:
This struct only holds the pointer to the key and the key size. It is used being set in one CC function  so I don't see an added value of putting it as part of the context structure.
In addition, if you have this structure as part of the context, which will only point to data being given by user, it might point to data that doesn't exist anymore

andresag01 replied:
ok

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+int mbedtls_cipher_cmac_update( mbedtls_cipher_context_t *ctx,
+                                const unsigned char *input, size_t ilen )
+{
+    mbedtls_cmac_context_t* cmac_ctx;
```
Style: Please note that the `*` usually goes next to the symbol name instead of the type

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+
+        /*
+         * Process the unproccessed data, in case it reached a full AES block,
+         * and there is there is still input data.
```
Typo in comment: `there is there is`

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+
+    if( ilen > 0 )
+    {
+        const size_t size_to_store = ( ilen % SASI_AES_BLOCK_SIZE_IN_BYTES == 0) ?
```
Style: Please add space before )

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+                                     SASI_AES_BLOCK_SIZE_IN_BYTES : ilen % SASI_AES_BLOCK_SIZE_IN_BYTES;
+        memcpy( &cmac_ctx->unprocessed_block[0],
+                           input + ilen - size_to_store,
+                           size_to_store );
```
Style: please fix alignment

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+    {
+        const size_t size_to_store = ( ilen % SASI_AES_BLOCK_SIZE_IN_BYTES == 0) ?
+                                     SASI_AES_BLOCK_SIZE_IN_BYTES : ilen % SASI_AES_BLOCK_SIZE_IN_BYTES;
+        memcpy( &cmac_ctx->unprocessed_block[0],
```
Is the address of element 0 necessary?

RonEld replied:
no, I think it was just a copy \ paste error from copying `memcpy( &cmac_ctx->unprocessed_block[cmac_ctx->unprocessed_len],`

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+        }
+    }
+
+    if( ilen > 0 )
```
I think the assumption for the code in this if statement is that there are is no data in unprocessed_block if `ilen == 0`. Could we write that down in a comment? I just think that the precondition is not obvious.

RonEld replied:
But this is not correct.
It is possible to have `ilen == 0` and data in `unprocessed_block`. This is in case the function is given an empty in buffer. However, in this case, if `unprocessed_block` is not full, we don't do anything with it, until `mbedtls_cipher_cmac_finish()` is called. 

In addition, it is possible that a non empty buffer is given as input, but it is very short ( e.g. 1 byte ), and the `unprocessed_block has not reached 16 bytes after storing the byte.

andresag01 replied:
Sorry, my comment was incorrect, I meant to say that the code in the if statement is executed only when unprocessed_block doesnt have any data.

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+        ilen -= size_to_store;
+        if( ilen > 0 )
+        {
+            if( SaSi_AesBlock( &cmac_ctx->CC_Context, ( uint8_t * )input,
```
Style: Please note that we do not normally add space before and after ) ( for type casts

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+        ilen -= size_to_store;
+        if( ilen > 0 )
+        {
+            if( SaSi_AesBlock( &cmac_ctx->CC_Context, ( uint8_t * )input,
```
What are the alignment requirements for `input` when calling this function? Is it necessary that the beginning be aligned to 2, 4, ..., SASI_AES_BLOCK_SIZE_IN_BYTES? I was trying to follow the code and I believe its possible to force a path in the code that ends up with `input` aligned to an arbitrary byte address. But I am not sure if this matters because I do not know the alignment requirements

RonEld replied:
It's not a matter of address alignment,AFAIK; it's a matter of size of input, that needs to be a multiple of 16 bytes, except for the finish operation

andresag01 replied:
Ok, so in summary: There are no alignment requirements. The only requirement is that the length of the input must be a multiple of the block size.

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+    {
+        mbedtls_platform_zeroize( cmac_ctx, sizeof( *cmac_ctx ) );
+        mbedtls_free( cmac_ctx );
+        SaSi_AesFree( &cmac_ctx->CC_Context );
```
Doesnt the order of these statements need to be changed? It looks to me like you are calling SaSi_AesFree() on an address within *cmac_ctx after freeing that memory

RonEld replied:
correct, this is abug

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+int mbedtls_cipher_cmac_finish( mbedtls_cipher_context_t *ctx,
+                                unsigned char *output )
+{
+    mbedtls_cmac_context_t* cmac_ctx;
```
Style: I think we normally put the `*` next to the symbol

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+        ret = init_cc( cmac_ctx );
+        if( ret != 0 )
+            goto exit;
+    }
```
Why is it that we need to call `init_cc()` here and in `mbedtls_cipher_cmac_update()` instead of calling it during initialization in `mbedtls_cipher_cmac_starts()`? It just looks to me like the program would not work without calling `mbedtls_cipher_cmac_starts()` at some point anyways because otherwise who would allocate the cmac context?

RonEld replied:
because in the flow, the intiailization can be called only once, and some sequential cmac calls can be separated with different calls to `mbedtls_cipher_cmac_starts()`.
`init_cc()` should be called to initialize a new CC CMAC context

andresag01 replied:
I understand the first part of your answer, but not this:
```
some sequential cmac calls can be separated with different calls to mbedtls_cipher_cmac_starts().
```
what do you mean?

RonEld replied:
I am not sure what I meant...
Anyway, I don't remember at the moment, but I believe I had a reason not call `init_cc()` from starts. I think it's because of the `reset` functionality.
But I think that in multithreading environment, we would like to set the key only before doing the actual operation.

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+
+int mbedtls_cipher_cmac_reset( mbedtls_cipher_context_t *ctx )
+{
+    mbedtls_cmac_context_t* cmac_ctx;
```
Style: Normally we put the `*` next to the symbol name rather than the type.

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+    return( ret );
+}
+
+int mbedtls_cipher_cmac_reset( mbedtls_cipher_context_t *ctx )
```
How is the sequence of events supposed to work if the user wants to reset the cmac? Is it something like:
```
starts
update
update 
...
finish
reset
starts
update
update
...
```

Or is it something like:
```
starts
update
update 
...
finish
reset
update
update
...
```
Does the second one even make sense? How would the user change the key (for example)? Also there would probably be a memory leak in that case because the cmac context will be reallocated.

RonEld replied:
According to [documentation for `mbedtls_cmac_reset()`](https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/cmac.h#L138):
~~~

 *                      It is called after mbedtls_cipher_cmac_finish()
 *                      and before mbedtls_cipher_cmac_update().
~~~
[mbedtls_cipher_cmac_finish()`](https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/cmac.h#L119):
~~~
 *                      It is called after mbedtls_cipher_cmac_update().
 *                      It can be followed by mbedtls_cipher_cmac_reset() and
 *                      mbedtls_cipher_cmac_update(), or mbedtls_cipher_free().
~~~
[`mbedtls_cipher_cmac_update()`](https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/cmac.h#L100):
~~~
*                      It is called between mbedtls_cipher_cmac_starts() or
 *                      mbedtls_cipher_cmac_reset(), and mbedtls_cipher_cmac_finish().
 *                      Can be called repeatedly.
~~~

andresag01 replied:
In other words, we do not need to call starts after reset.

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+ * \return              \c 0 on success.
+ * \return              #MBEDTLS_ERR_MD_BAD_INPUT_DATA
+ *                      if parameter verification fails.
+ */
```
Why do we have this documentation here? Surely its the same as in the header files. But probably nobody is going to see it as its inside this source file.

RonEld replied:
A mistake. I will remove

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+    }
+
+exit:
+    if( SaSi_AesFree( &cmac_ctx->CC_Context ) != 0 && ret == 0 )
```
Does this function need to be called somewhere else? What happens if I am the application programmer and decide that somewhere in the middle of a computation I no longer want to continue doing CMAC (even though all the calls to starts and update have terminated correctly). How do I release the resources from that CMAC context and the accelerator?

RonEld replied:
Since there isn't any `terminate` function to deallocate all of the cmac resources, I don't see another location

andresag01 replied:
Well, but how does this API work then? how does the user release the resources from a context? What happens when I decided that I do not want to continue a computation? Who releases the dynamic memory allocated?

RonEld replied:
To be honest, I don't have a good answer, but I really don't see another option of where to clear the resources, the way our `cmac` is implemented.
The cmac context is cleared in the `cipher` module, which I don't think is correct, but this is the way it behaves.
Anyway, as mentioned in mbedtls_cipher_free , this is not a real memory leak, as the CC function only clears the context,

Patater replied:
To free the resources, `reset` should be called on the context. `SaSi_AesFree` should be called as part of `mbedtls_cmac_reset()`

RonEld replied:
But if you look at implementation of `mbedtls_cipher_cmac()`, you will see that `mbedtls_cmac_reset()` is not called, causing resources not to be freed

Patater replied:
Sounds like a bug

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+
+
+exit:
+    if( SaSi_AesFree( &ctx.cmac_ctx->CC_Context ) != 0 && ret == 0 )
```
What happens if we fail before calling `init_cc`? Is that ok/acceptable?

RonEld replied:
Nothing will really happen, because `SaSi_AesFree()` mostly memsets to zero. However I see your point, as this is not correct behavior. i'll fix this

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.h:
```diff
@@ -0,0 +1,50 @@
+/*
+ *  aes_alt.h
+ *
+ *  Copyright (C) 2018, ARM Limited, All Rights Reserved
```
Note that the dates and branding of this license header are out-of-date

RonEld replied:
yes, copy\paste error as well

andresag01 commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,333 @@
+    cmac_ctx->unprocessed_len = 0;
+    mbedtls_platform_zeroize( cmac_ctx->unprocessed_block,
+                              sizeof( cmac_ctx->unprocessed_block ) );
+
```
Why doesnt this function need to reset the accelerator and clear the `initialized` flag.

RonEld replied:
[because](https://github.com/ARMmbed/mbedtls/blob/development/include/mbedtls/cmac.h#L134):
~~~
 * \brief               This function prepares the authentication of another
 *                      message with the same key as the previous CMAC
 *                      operation.
~~~
Since same key is used, no need to reset the accelerator

andresag01 replied:
ok

andresag01 reviewed (changes requested):
@RonEld: I did an initial review of the changes and wrote down a number of comments/questions. Please let me know what you think

Commit:
Style fixes

Fix minor style fixes and typos:
1. Change file name to correct one.
2. Change copyright year.
3. Insert whitespaces before and after paranthesis.
4. Put `*` next to pointer name.

Commit:
Fix functionality issues

FIx some functionality issues for better visibility:
1. Allocate the context only for 128 bit key
2. Change oreder of freeing the resources.

Commit:
Make the alternative cmac optional

Have the alternative cmac undefined by default,
in order not to break backwards compatability.

Patater commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,303 @@
+    mbedtls_platform_zeroize( cmac_ctx->unprocessed_block,
+                              sizeof( cmac_ctx->unprocessed_block ) );
+
+    return( 0 );
```
`reset()` should also call `SaSi_AesFree`, to free up the `cmac_ctx->CC_Context`. I'm not sure if we need to check that CC_Context was initialized (as we initialize it in `update()`) before we reset. Alternatively, we could consider allocating the CC_Context in `starts()`.

RonEld replied:
I thought about allocating in `starts()`, but then in use cases of `starts() ->update()->finish() ->reset() update()->finish()` we might fail

Patater replied:
So, both starts and reset would need to allocate a new CC_Context context as well in that flow. But, reset is supposed to free all resources, not make new ones. This means starts is not a good place to allocate resources, agreed.

Patater commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.h:
```diff
@@ -0,0 +1,50 @@
+extern "C" {
+#endif
+
+struct mbedtls_cmac_context_t
```
Do we need to typedef this so it works with C? It looks like perhaps this header was only tested with C++ builds.

RonEld replied:
This was tested with our on target tests, which I believe is in c

Patater replied:
Mbed OS always builds with C++, so it's always C++ for on-target testing. We should use typedef here if we want to be C compatible. Otherwise, when you later refer to `mbedtls_cmac_context_t blah;` instead of `struct mbedtls_cmac_context_t blah`, you'll get compilation failures. I see many places in this PR where we don't say `struct mbedtls_cmac_context_t`.

RonEld replied:
sure, I'll change but this implementation is for Mbed OS anyway

Commit:
Handle CC context correct

Initiate the CC context in the starts function and in the reset.
In the reset function, free aes context before.
Free the context in the finish function and reset function.

Commit:
Make the cmac context a typedef

Make the cmac context a typedef, to be compatible with c code.

Patater commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,292 @@
+
+    if( ctx == NULL || ctx->cipher_info == NULL )
+        return( MBEDTLS_ERR_CIPHER_BAD_INPUT_DATA );
+
```
Remove extra newline

Patater commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,292 @@
+exit:
+    if( ret != 0 )
+    {
+        SaSi_AesFree( &cmac_ctx->CC_Context );
```
Is it always safe to call `SaSi_AesFree()`? or do we need to check `cmac_ctx->is_cc_initiated == 1` like we do in `mbedtls_cipher_cmac_finish()`? You might want to make `deinit_cc()` to avoid repeating the same `is_cc_initiated` check everywhere.

RonEld replied:
>Is it always safe to call SaSi_AesFree()?

Well, yes, because internally it just memsets the buffer to zero, but It's better to change as you suggested. It's a residue from when we initiated the context also in update

Patater commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,292 @@
+    ret = mbedtls_cipher_cmac_starts( &ctx, key, keylen );
+    if( ret != 0 )
+        goto exit;
+
```
Remove extra newline.

Commit:
Add a deinit function

Add a deinit function that will be called and check inside
whether context is initialized. This function is called for
freeing the CC context, instead of every time check that it's
initizliaed and free it.

Patater reviewed (approved):
LGTM

Please comment what testing you've done, now that this has undergone some changes during code review.

Patater reviewed (approved):
Thanks. LGTM

SeppoTakalo commented on features/cryptocell/FEATURE_CRYPTOCELL310/cmac_alt.c:
```diff
@@ -0,0 +1,294 @@
+                      SASI_AES_MODE_CMAC, SASI_AES_PADDING_NONE ) != 0 )
+    {
+        return( MBEDTLS_ERR_PLATFORM_HW_ACCEL_FAILED );
+    }
```
Not sure why Astyle does not complain, but this is not really Mbed Coding style.
We should follow more K&R style. See https://os.mbed.com/docs/mbed-os/v5.13/contributing/style.html

gilles-peskine-arm replied:
This is an alternative implementation for code in Mbed TLS, which has its own style rules. Mbed TLS and alternative implementations for it are deliberately excluded from the astyle check in Mbed OS. I don't know if there's a formal rule stated somewhere, but most alternative implementations follow the Mbed TLS style.
